# Aula 3 - Loop While

Na aula de hoje, vamos explorar o seguinte tópico em Python:

- 1) Laços de repetição (while).

_______

### Objetivos

Laço de repetição while, frisando aqui a necessidade de atualização da condição para evitar loops infinitos. Nesta aula, será mostrado como os operadores lógicos são utilizados nas novas estruturas apresentadas. 

____
____
____

## 3) Laços de repetição (while)

Uma das utilidades de linguagens de programação é a de automatizar tarefas que são repetitivas.

Mas, pra isso ser viável, seria bom se tivéssemos uma estrutura para **repetir comandos**, não é mesmo?

Imagine que eu queira exibir na tela "Olá, mundo!" 5 vezes. Podemos fazer:

In [1]:
print("Olá, mundo!")
print("Olá, mundo!")
print("Olá, mundo!")
print("Olá, mundo!")
print("Olá, mundo!")

Olá, mundo!
Olá, mundo!
Olá, mundo!
Olá, mundo!
Olá, mundo!


In [5]:
contador = 0
while contador < 5:
    print('Olá, mundo!')
    contador += 1

Olá, mundo!
Olá, mundo!
Olá, mundo!
Olá, mundo!
Olá, mundo!


Mas, e se eu quiser exibir essa mensagem 1000 vezes? Ou 1 milhão de vezes? Não é ideal escrevermos o mesmo pedaço de código tantas vezes, né?

Para isso, existem os **laços de repetição**, que permitem repetir pedaços de código quantas vezes desejarmos!

O primeiro laço que vamos ver é o **while**. Este laço tem a seguinte estrutura:

```python
while (condicao é True):
    operacao_repetida
```

Ou seja, o que tá no bloco do while é repetido **enquanto a condição for verdadeira**

Isso pode levar a **loops infinitos**

In [ ]:
while 2==2:
    print("Olá, mundo")

Para que loops infinitos não aconteçam, temos que fazer uma **atualização da condição** a cada iteração do laço!

Isso é, temos que **atualizar** a variavel que contabiliza as repetições no loop

In [ ]:
# vamos definir uma variável
# "x recebe o valor'aa' 2"

x = 2

# podemos redefinir a variável em termos do valor anterior dela
# "x recebe o valor que tinha ANTES na variável, + 3"
x = x + 3

# chamando a variável
x

Assim, o que fazemos é **definir a condição do while em termos de uma variável que tenha seu valor atualizado!**

Para isso, é comum nos referirmos à variàvel da condição como **variável contadora**.

Para atualizar a variável contadora dentro do while, em geral a atualizamos em +1

In [7]:
cont = 0

while cont < 5:
    print("Olá mundo")
    cont += 1
    print(cont)
    print(cont < 5)

Olá mundo
1
True
Olá mundo
2
True
Olá mundo
3
True
Olá mundo
4
True
Olá mundo
5
False


Vamos entender um pouco melhor como a variável contadora se comporta?

Pra isso, basta exibi-la a cada iteração:

In [ ]:
cont = 0

while cont < 5:
    
    print(cont)
    
    # cont += 1
    cont = cont + 1

O código acima equivale a:

In [ ]:
# inicialmente, a variavel cont vale 0
cont = 0
print(cont)

# pegamos o valor anterior (0) e somamos 1, ficando 1
cont = cont + 1
print(cont)

# pegamos o valor anterior (1) e somamos 1, ficando 2
cont = cont + 1
print(cont)

# pegamos o valor anterior (2) e somamos 1, ficando 3
cont = cont + 1
print(cont)

# pegamos o valor anterior (3) e somamos 1, ficando 4
cont = cont + 1
print(cont)

Usar a condição com < ao invés de != em geral garante maior segurança ao seu algoritmo, pois evita imprevistos caso você receba números do usuário!

Podemos também atualizar a condição de repetição segundo informado pelo usuário

Esse uso é bem importante para **garantir que o usuário digitou corretamente o que foi solicitado**

Por exemplo, vamos pedir pro usuário digitar um número maior que 10. **Enquanto ele não fizer o que queremos**, vamos continuar pedindo pra ele digitar um novo valor:

In [12]:
contador = 0
while contador < 5:
    print(contador)
    contador += 2

0
2
4


In [13]:
numero = 0

while numero <= 10:
    numero = int(input("Digite um numero maior que 10: "))
    
print("O numero digitado foi: ", numero)

O numero digitado foi:  15


Outro exemplo, onde pedirmos pro usuário digitar sua nota (para ser um valor válido, tem que estar entre 0 e 10!)

In [14]:
nota = -1

while not(0 <= nota <= 10):
    nota = int(input("Digite um número de 0 a 10: "))
    
print("Nota: ", nota)

Nota:  8


Outro exemplo...

In [16]:
genero = ""

while genero != "M" and genero != "F" and genero != "O":
    genero = input("Digite seu genero: ")
    
print("Genero: ", genero)

Genero:  O


In [19]:
tipoCombustivel = input("Qual o tipo de combustivel desejado (g- gasolina/a - alcool): ").upper()

while tipoCombustivel != "A" and tipoCombustivel != "G":
    print("Tipo inválido!")
    tipoCombustivel = input("Qual o tipo de combustivel desejado (g- gasolina/a - alcool): ")
    
print("Tipo:", tipoCombustivel)

Tipo: G


## break, continue e else

O while pode ser parado no meio com o break, isso é utilizado caso exista algum(ns) casos excepcionais onde seu código deve parar ou mesmo para poupar processamento em um loop que já fez o que precisava e usava a condição inicial apenas como limite de iterações.

In [20]:
cont = 0

while cont < 10:
    cont += 1
    if cont == 3:
        break
    print(cont)

1
2


Com o comando continue conseguimos parar uma iteração no meio e recomeçar o bloco do while.

In [23]:
cont = 0

while cont < 10:
    cont += 1
    if cont == 3:
        continue
    print(cont)

1
2
4
5
6
7
8
9
10


O else aqui vai se comportar como um **"e então"**, pois será o bloco de código que rodará quando, se **e somente se**, a condição do while não for mais satisfeita.

In [25]:
cont = 0

while cont < 10:
    cont += 1
    if cont == 5:
        break
    print(cont)
else:
    print("Fim")

1
2
3
4


# Calculando primos com loops

In [29]:
numero = int(input("Digite o numero que quer validar: "))
divisor = 2

while divisor < (numero/2)+1:
    if numero % divisor == 0:
        print("Não é primo")
        break
    divisor += 1
else:
    print("É primo")

Não é primo


In [31]:
%%timeit

numero = 10000000
divisor = 2

while divisor < (numero/2)+1:
    if numero % divisor == 0:
        break
    divisor += 1
else:
    pass

247 ns ± 8.61 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [32]:
%%timeit

numero = 10000000
divisor = 2

while divisor < (numero/2)+1:
    if numero % divisor == 0:
        pass
    divisor += 1
else:
    pass

1.2 s ± 32.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


---

## Exercícios

**1)** Print the following patterns using loop :

a.<br>

\*<br>
\*\*<br>
\*\*\*<br>
\*\*\*\*<br>

b.<br>

  \*  <br>
 \*\*\* <br>
\*\*\*\*\*<br>
 \*\*\* <br>
  \*  <br>

**2)** Reverse a given integer number

**3)** Escreva um programa que encontra o maximo divisor comum entre dois números

**4)** Faça a somatória de todos os numeros seguindo a sequência: <br>

$ \frac{1}{1} + \frac{1}{2} + \frac{1}{3} + \frac{1}{4} + \frac{1}{5} + \frac{1}{6} + ... + \frac{1}{1000} $




$ \frac{1}{1} + \frac{1}{2} + \frac{1}{4} + \frac{1}{8} + \frac{1}{16} + \frac{1}{32} + ... + \frac{1}{2^{n}} $



**5)** Seguindo a mesma ideia do exercício 4, faça a soma da seguinte sequência: <br>

$ \frac{1}{1!} + \frac{1}{2!} + \frac{1}{3!} + \frac{1}{4!} + \frac{1}{5!} + \frac{1}{6!} + ... + \frac{1}{100!} $

**6)** Faça uma calculadora de investimentos que define se um fundo gera lucro real (rendimento acima da inflação) ao final de 12 meses, considerando o rendimento mensal projetado, a taxa de administração do fundo, a inflação no ano, come-cotas ( vamos considerar 15%) e o capital investido. Caso positivo, informe quanto foi esse lucro.